access/access-impl/impl/src/java/org/sakaiproject/access/tool/AccessServlet.java

```diff
@@ -346,6 +357,55 @@ public SecurityAdvice isAllowed(String userId, String function, String reference
 			HttpAccess access = service.getHttpAccess();
 			if (access == null) throw new EntityNotDefinedException(ref.getReference());
 
+			/** SAK-45537 Content: Ability to consume resources using time based tokens. */
+			/*
+			 * Time based tokens can be generated from different tools with the user's sessionId.
+			 * If the current time is before the limit date of the file and the session is valid, we should provide access to the file no matter the permissions.
+			 * Tokens have this format XXXXXXXXXXXXXXX|2021-05-25T15:38:00.955106300
+			 * XXXXXXXXXXXXXXX is the session that must match with the current session
+			 * 2021-05-25T15:38:00.955106300 is the validity date of the token, generated by the producer.
+			 */
+			String secureTokenParameter = req.getParameter(SECURE_FILE_TOKEN_PARAM);
+			// Check if the securetoken param has been sent from the client.
+			if (StringUtils.isNotBlank(secureTokenParameter)) {
+				try {
+					log.debug("The secure token parameter has been received {}", secureTokenParameter);
+					String decryptedToken = encryptionUtilityService.decrypt(secureTokenParameter);
+					String[] tokenParts = decryptedToken.split("\\|");
+					String decryptedSessionValue = tokenParts[0];
+					String decryptedDateValue = tokenParts[1];
+					log.debug("The decryptedSessionValue is {}", decryptedSessionValue);
+					log.debug("The decryptedDateValue is {}", decryptedDateValue);
+					LocalDateTime decryptedTokenValidity = LocalDateTime.parse(decryptedDateValue);
+					boolean isValidSession = sessionManager.getCurrentSession().getId().equals(decryptedSessionValue);
+					boolean isValidTime = LocalDateTime.now().isBefore(decryptedTokenValidity);
+					// If the time-based token is correct and the file contains the secured property, serve the file.
+					if (isValidSession && isValidTime) {
+						log.debug("The token {} is valid for the ref {}", secureTokenParameter, ref);
+						// get the properties - but use a security advisor to avoid needing end-user permission to the resource
+						SecurityAdvisor securityAdvisor = new SecurityAdvisor() {
+							public SecurityAdvice isAllowed(String userId, String function, String reference) {
+								return SecurityAdvice.ALLOWED;
+							}
+						};
+						try {
+							securityService.pushAdvisor(securityAdvisor);
+							ResourceProperties props = ref.getProperties();
+							if (props.get(ResourceProperties.PROP_SECURED) != null) {
+								access.handleAccess(req, res, ref, accepted);
+								return;
+							}
+						} catch (Exception e) {
+							log.error("Error serving a secured resource {}", e.getMessage());
+						} finally {
+							securityService.popAdvisor(securityAdvisor);
+						}
+					}
+				} catch (Exception e) {
+					// Do not reveal failed requests.
```

Comment:
how about a debug log for people testing out?

log.debug("error while decrypting secure token", e)

Reply:
Thanks @ottenhoff, I've included it.